cub/test/catch2_test_device_merge_sort.cu

```diff
@@ -0,0 +1,318 @@
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#include <cub/device/device_merge_sort.cuh>
+
+#include <thrust/copy.h>
+#include <thrust/detail/raw_pointer_cast.h>
+#include <thrust/device_vector.h>
+#include <thrust/fill.h>
+#include <thrust/host_vector.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/random.h>
+#include <thrust/sequence.h>
+#include <thrust/shuffle.h>
+
+#include <algorithm>
+
+#include "catch2_test_device_merge_sort_common.cuh"
+#include "catch2_test_helper.h"
+#include "catch2_test_launch_helper.h"
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortPairs, sort_pairs);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortPairsCopy, sort_pairs_copy);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::StableSortPairs, stable_sort_pairs);
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortKeys, sort_keys);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortKeysCopy, sort_keys_copy);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::StableSortKeys, stable_sort_keys);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::StableSortKeysCopy, stable_sort_keys_copy);
+
+using key_types =
+  c2h::type_list<std::uint8_t,
+                 std::int16_t,
+                 std::uint32_t,
+                 double,
+                 c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>>;
+using wide_key_types = c2h::type_list<std::uint32_t, double>;
+
+using value_types =
+  c2h::type_list<std::uint8_t, float, c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>>;
+
+/**
+ * Function object that maps a the targeted sorted rank of an item to a key.
+
+ * E.g., `OffsetT` is `int32_t` and `KeyT` is `float`:
+ * [  4,   2,   3,   1,   0] <= targeted key ranks
+ * [4.0, 2.0, 3.0, 1.0, 0.0] <= corresponding keys
+ */
+template <typename OffsetT, typename KeyT>
+struct rank_to_key_op_t
+{
+  __device__ __host__ KeyT operator()(const OffsetT& val)
+  {
+    return static_cast<KeyT>(val);
+  }
+};
+
+template <typename OffsetT>
+struct rank_to_key_op_t<OffsetT, c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>>
+{
+  using custom_t = c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>;
+  __device__ __host__ custom_t operator()(const OffsetT& val)
+  {
+    custom_t custom_val{};
+    custom_val.key = val;
+    custom_val.val = val;
+    return custom_val;
+  }
+};
+
+/**
+ * Generates a shuffled array of key ranks. E.g., for a vector of size 5: [4, 2, 3, 1, 0]
+ */
+template <typename OffsetT>
+void get_key_ranks(thrust::device_vector<OffsetT>& key_ranks, c2h::seed_t seed)
```

Comment:
This function is only ever called directly after the construction of `key_ranks`

Can we move that construction into this function, so that it takes `num_items` as an argument and returns a `device_vector`?